File: C++/Swiat.cpp
#include "Swiat.h"
#include <windows.h>
#include <fstream>
#include "Czlowiek.h"
#include "Wilk.h"
#include "Owca.h"
#include "Lis.h"
#include "Zolw.h"
#include "Antylopa.h"
#include "Trawa.h"
#include "Mlecz.h"
#include "Guarana.h"
#include "WilczeJagody.h"
#include "BarszczSosnowskiego.h"

Swiat::Swiat()
{
	this->organizmy = new Organizm * [40 * 40];

	for (int i = 0; i < 40*40; i++)
	{
		organizmy[i] = nullptr;
	}

	this->szerokosc = 40;
	this->wysokosc = 40;
	this->Setkierunek(BRAK);
	this->aktywna = false;
	this->czlowiekzyje = true;
	this->rundy = CZAS_UMIEJETNOSCI;
	this->komunikat = "";
}

Swiat::Swiat(int m)
{
	this->organizmy = new Organizm * [m * 40];

	for (int i = 0; i < m*40; i++)
	{
		organizmy[i] = nullptr;
	}

	this->szerokosc = m;
	this->wysokosc = 40;
	this->Setkierunek(BRAK);
	this->aktywna = false;
	this->czlowiekzyje = true;
	this->rundy = CZAS_UMIEJETNOSCI;
	this->komunikat = "";
}

Swiat::Swiat(int m, int n)
{
	this->organizmy = new Organizm * [m * n];

	for (int i = 0; i < m * n; i++)
	{
		organizmy[i] = nullptr;
	}
	
	this->szerokosc = m;
	this->wysokosc = n;
	this->Setkierunek(BRAK);
	this->aktywna = false;
	this->czlowiekzyje = true;
	this->rundy = CZAS_UMIEJETNOSCI;
	this->komunikat = "";

	Organizm* org = new Czlowiek(0, 0, this);
	int id = 0, x, y;
	for (int i = 0; i < (szerokosc * wysokosc) / 15; i++)
	{
		id = rand() % 11 + 2;
		x = rand() % szerokosc;
		y = rand() % wysokosc;
		if (this->GetOrganizmy(x, y) == nullptr)
		{
			if (id == IDWILK)
				org = new Wilk(x, y, this);
			else if (id == IDOWCA)
				org = new Owca(x, y, this);
			else if (id == IDLIS)
				org = new Lis(x, y, this);
			else if (id == IDZOLW)
				org = new Zolw(x, y, this);
			else if (id == IDANTYLOPA)
				org = new Antylopa(x, y, this);
			else if (id == IDTRAWA)
				org = new Trawa(x, y, this);
			else if (id == IDMLECZ)
				org = new Mlecz(x, y, this);
			else if (id == IDGUARANA)
				org = new Guarana(x, y, this);
			else if (id == IDWILCZEJAGODY)
				org = new WilczeJagody(x, y, this);
			else if (id == IDBARSZCZSOS)
				org = new BarszczSosnowskiego(x, y, this);
		}

	}


}
void Swiat::wykonajTure()
{
	for (int i = ILIST - 1; i >= 0; i--)
	{
		int j = 0;
		while (j < kolejnosc[i].size())
		{
			if (kolejnosc[i][j]->GetDZ() == false)
			{
				kolejnosc[i][j]->KolejnaRunda();
				kolejnosc[i][j]->akcja();
			}
			else
				kolejnosc[i][j]->SetDZ(false);
			j++;
		}
	}
	if (this->rundy >= CZAS_UMIEJETNOSCI && this->aktywna == true)
	{
		this->ZmienAktywna();
		this->WyzerujRundy();
		this->dodajWydarzenie(nullptr, nullptr, UMIEJETNOSC_N);
	}
	this->KolejnaRunda();
}

void Swiat::umiejetnosc()
{
	if(this->czlowiekzyje == false)
	{
		this->komunikat = "Nie mozna uzyc umiejetnosci, czlowiek nie zyje";
	}
	else if (this->rundy >= CZAS_UMIEJETNOSCI)
	{
		this->ZmienAktywna();
		this->WyzerujRundy();
		if(this->aktywna)
			this->dodajWydarzenie(nullptr, nullptr, UMIEJETNOSC_A);
		else
			this->dodajWydarzenie(nullptr, nullptr, UMIEJETNOSC_N);
	}
	else
	{
		if (this->CzyAktywna())
		{
			//juz aktywna ile tur jeszcze
			this->komunikat = "Umiejetnosc aktywna przez ";
			this->komunikat += char(CZAS_UMIEJETNOSCI - this->rundy + 48);
			if (CZAS_UMIEJETNOSCI - this->rundy == 1)
				this->komunikat += " ture";
			else if (CZAS_UMIEJETNOSCI - this->rundy == 5)
				this->komunikat += " tur";
			else
				this->komunikat += " tury";
		}
		else
		{
			//umiejetnosc gotowa w nastepnej turze
			this->komunikat = "Umiejetnosc gotowa za ";
			this->komunikat += char(CZAS_UMIEJETNOSCI - this->rundy + 48);
			if (CZAS_UMIEJETNOSCI - this->rundy == 1)
				this->komunikat += " ture";
			else if (CZAS_UMIEJETNOSCI - this->rundy == 5)
				this->komunikat += " tur";
			else
				this->komunikat += " tury";
		}
	}
}

void Swiat::rysujSwiat()
{
	HANDLE col;
	col = GetStdHandle(STD_OUTPUT_HANDLE);
	SetConsoleTextAttribute(col, KOLORDEFAULT);
	int i;
	for (int j = 0; j < (szerokosc + 1) * 2; j++)
		std::cout << "-";
	std::cout << "\n";
	for (i = 0; i < wysokosc; i++)
	{
		SetConsoleTextAttribute(col, KOLORDEFAULT);
		std::cout << "|";
		for (int j = 0; j < szerokosc; j++)
		{
			if (organizmy[i * szerokosc + j] != nullptr)
				organizmy[i * szerokosc + j]->rysowanie();
			else
			{
				std::cout << " ";
			}
			std::cout << " ";
		}
		SetConsoleTextAttribute(col, KOLORDEFAULT);
		std::cout << "| ";
		if (this->wydarzenia.size() > i)
			std::cout << this->wydarzenia[i];
		std::cout << "\n";
	}
	col = GetStdHandle(STD_OUTPUT_HANDLE);
	SetConsoleTextAttribute(col, KOLORDEFAULT);
	for (int j = 0; j < (szerokosc+1)*2; j++)
		std::cout << "-";

	std::cout << "\nVincenzo Piras 193218\nLegenda: \nstrzalki - ruch, spacja - umiejetnosc, f - zakoncz program, s - zapisz do pliku, l - wczytaj z pliku\n";
	SetConsoleTextAttribute(col, KOLORCZLOWIEK);
	std::cout << "H";
	SetConsoleTextAttribute(col, KOLORDEFAULT);
	std::cout << " - czlowiek, ";
	SetConsoleTextAttribute(col, KOLORWILK);
	std::cout << "W";
	SetConsoleTextAttribute(col, KOLORDEFAULT);
	std::cout << " - wilk, ";
	SetConsoleTextAttribute(col, KOLOROWCA);
	std::cout << "O";
	SetConsoleTextAttribute(col, KOLORDEFAULT);
	std::cout << " - owca, ";
	SetConsoleTextAttribute(col, KOLORLIS);
	std::cout << "L";
	SetConsoleTextAttribute(col, KOLORDEFAULT);
	std::cout << " - lis\n";
	SetConsoleTextAttribute(col, KOLORZOLW);
	std::cout << "Z";
	SetConsoleTextAttribute(col, KOLORDEFAULT);
	std::cout << " - zolw, ";
	SetConsoleTextAttribute(col, KOLORANTYLOPA);
	std::cout << "A";
	SetConsoleTextAttribute(col, KOLORDEFAULT);
	std::cout << " - antylopa, ";
	SetConsoleTextAttribute(col, KOLORTRAWA);
	std::cout << "T";
	SetConsoleTextAttribute(col, KOLORDEFAULT);
	std::cout << " - trawa, ";
	SetConsoleTextAttribute(col, KOLORMLECZ);
	std::cout << "M";
	SetConsoleTextAttribute(col, KOLORDEFAULT);
	std::cout << " - mlecz\n";
	SetConsoleTextAttribute(col, KOLORGUARANA);
	std::cout << "G";
	SetConsoleTextAttribute(col, KOLORDEFAULT);
	std::cout << " - guarana, ";
	SetConsoleTextAttribute(col, KOLORWILCZEJAGODY);
	std::cout << "J";
	SetConsoleTextAttribute(col, KOLORDEFAULT);
	std::cout << " - wilcze jagody, ";
	SetConsoleTextAttribute(col, KOLORBARSZCZSOS);
	std::cout << "B";
	SetConsoleTextAttribute(col, KOLORDEFAULT);
	std::cout << " - barszcz sosnowskiego\n";

	std::cout << this->komunikat << "\n";
	this->komunikat = "";

	while (i < this->wydarzenia.size())
	{
		std::cout << wydarzenia[i];
		i++;
	}

	this->wydarzenia.clear();

}

void Swiat::addToList(Organizm* organizm)
{
	int i = organizm->GetInicjatywa();
	if (i < 2)
	{
		kolejnosc[i].push_back(organizm);
	}
	else if (i == 5)
	{
		kolejnosc[3].push_back(organizm);
	}
	else if (i == 7)
	{
		kolejnosc[4].push_back(organizm);
	}
	else
	{
		kolejnosc[2].push_back(organizm);
	}
}

void Swiat::deletefromList(Organizm* organizm)
{
	int i = organizm->GetInicjatywa();
	int j = 0;

	Czlowiek* test = dynamic_cast<Czlowiek*>(organizm);
	if (test != nullptr)
		this->czlowiekzyje = false;

	if (i < 2)
	{
		while (j < kolejnosc[i].size())
		{
			if (kolejnosc[i][j] != organizm)
				j++;
			else
				kolejnosc[i].erase(kolejnosc[i].begin() + j);
		}
	}
	else if (i == 5)
	{
		while (j < kolejnosc[3].size())
		{
			if (kolejnosc[3][j] != organizm)
				j++;
			else
				kolejnosc[3].erase(kolejnosc[3].begin() + j);
		}
	}
	else if (i == 7)
	{
		while (j < kolejnosc[4].size())
		{
			if (kolejnosc[4][j] != organizm)
				j++;
			else
				kolejnosc[4].erase(kolejnosc[4].begin() + j);
		}
	}
	else
	{
		while (j < kolejnosc[2].size())
		{
			if (kolejnosc[2][j] != organizm)
				j++;
			else
				kolejnosc[2].erase(kolejnosc[2].begin() + j);
		}
	}
}
int Swiat::Getszerokosc() const
{
	return szerokosc;
}
int Swiat::Getwyskosc() const
{
	return wysokosc;
}

void Swiat::Setkierunek(int kierunek)
{
	this->kierunek_ruchu = kierunek;
}
int Swiat::Getkierunek() const
{
	return this->kierunek_ruchu;
}

void Swiat::SetOrganizmy(int x, int y, Organizm* n)
{
	this->organizmy[y * this->Getszerokosc() + x] = n;
}
Organizm* Swiat::GetOrganizmy(int x, int y) const
{
	return this->organizmy[y* this->Getszerokosc() + x];
}

void Swiat::WyzerujRundy()
{
	this->rundy = 0;
}
void Swiat::KolejnaRunda()
{
	this->rundy++;
}
int Swiat::GetRundy()
{
	return this->rundy;
}

void Swiat::SetRundy(int r)
{
	this->rundy = r;
}

bool Swiat::CzyAktywna()
{
	return this->aktywna;
}
void Swiat::ZmienAktywna()
{
	if (this->aktywna)
		this->aktywna = false;
	else
		this->aktywna = true;
}


void Swiat::dodajWydarzenie(Organizm* organizm1, Organizm* organizm2, int kod)
{
	std::string s = "";

	if (organizm1 != nullptr)
		s += organizm1->UczestnikWydarzenia();

	if (kod == ROZMNAZANIE)
		s += " rozmnazyl sie z ";
	else if (kod == ROZPRZESTRZENIANIE)
		s += " rozprzestrzenil sie";
	else if (kod == KILL)
		s += " zjadl ";
	else if (kod == ODBITY)
		s += " odbil atak ";
	else if (kod == UCIEKA)
		s += " ucieka od ";
	else if (kod == UMIEJETNOSC_A)
		s += "Czlowiek uzyl umiejetnosci";
	else if (kod == UMIEJETNOSC_N)
		s += "Umiejetnosc przestala dzialac";

	if (organizm2 != nullptr)
		s += organizm2->UczestnikWydarzenia();

	this->wydarzenia.push_back(s);
}

void Swiat::save()
{
	FILE* plik;
	fopen_s(&plik, "save.bin", "w");
	if (plik != NULL)
	{
		/*
		std::vector < Organizm* > kolejnosc[ILIST];
		std::vector < std::string > wydarzenia;
		int wysokosc;
		int szerokosc;
		int kierunek_ruchu;
		int rundy;
		bool aktywna;
		bool czlowiekzyje;
		std::string komunikat;
		Organizm** organizmy;
		*/
		int d = this->wysokosc;
		fwrite(&d, sizeof(int), 1, plik);
		d = this->szerokosc;
		fwrite(&d, sizeof(int), 1, plik);
		d = this->kierunek_ruchu;
		fwrite(&d, sizeof(int), 1, plik);
		d = this->rundy;
		fwrite(&d, sizeof(int), 1, plik);
		bool b = this->aktywna;
		fwrite(&b, sizeof(bool), 1, plik);
		b = this->czlowiekzyje;
		fwrite(&b, sizeof(bool), 1, plik);

		int size = 0;
		for (int i = 0; i < ILIST; i++)
			size += kolejnosc[i].size();

		fwrite(&size, sizeof(int), 1, plik);

		fclose(plik);
		//fopen_s(&plik, "org.bin", "w");
		std::ofstream outfile;
		outfile.open("output.txt", std::ios::out);
		if (outfile) {
			for (int i = ILIST - 1; i >= 0; i--)
			{
				int j = 0;
				while (j < kolejnosc[i].size())
				{
					/*
					int tmp = kolejnosc[i][j]->GetId();
					fwrite(&tmp, sizeof(int), 1, plik);
					tmp = kolejnosc[i][j]->GetX();
					fwrite(&tmp, sizeof(int), 1, plik);
					tmp = kolejnosc[i][j]->GetY();
					fwrite(&tmp, sizeof(int), 1, plik);
					tmp = kolejnosc[i][j]->GetSila();
					fwrite(&tmp, sizeof(int), 1, plik);
					tmp = kolejnosc[i][j]->GetRundy();
					fwrite(&tmp, sizeof(int), 1, plik);
					bool tmp2 = kolejnosc[i][j]->GetDZ();
					fwrite(&tmp2, sizeof(bool), 1, plik);
					*/

					int tmp = kolejnosc[i][j]->GetId();
					outfile << tmp << " ";
					tmp = kolejnosc[i][j]->GetX();
					outfile << tmp << " ";
					tmp = kolejnosc[i][j]->GetY();
					outfile << tmp << " ";
					tmp = kolejnosc[i][j]->GetSila();
					outfile << tmp << " ";
					tmp = kolejnosc[i][j]->GetRundy();
					outfile << tmp << " ";
					bool tmp2 = kolejnosc[i][j]->GetDZ();
					outfile << tmp2 << " ";

					j++;
				}
			}
		}
		outfile.close();
		//fclose(plik);
	}


}
void Swiat::load()
{
	FILE* plik;
	fopen_s(&plik, "save.bin", "r");
	if (plik != NULL)
	{
		for (int i = 0; i < szerokosc * wysokosc; i++)
		{
			if (organizmy[i] != nullptr)
				delete organizmy[i];
		}

		delete[] organizmy;


		for (int i = 0; i < ILIST; i++)
			this->kolejnosc[i].clear();


		int szerokoscr, wysokoscr, kierunek_ruchur, rundyr, size;
		bool aktywnar, czlowiekzyjer;

		fread(&wysokoscr, sizeof(int), 1, plik);
		fread(&szerokoscr, sizeof(int), 1, plik);
		fread(&kierunek_ruchur, sizeof(int), 1, plik);
		fread(&rundyr, sizeof(int), 1, plik);
		fread(&aktywnar, sizeof(bool), 1, plik);
		fread(&czlowiekzyjer, sizeof(bool), 1, plik);
		fread(&size, sizeof(int), 1, plik);

		fclose(plik);


		this->wysokosc = wysokoscr;
		this->szerokosc = szerokoscr;
		this->kierunek_ruchu = kierunek_ruchur;
		this->rundy = rundyr;
		this->aktywna = aktywna;
		this->czlowiekzyje = czlowiekzyjer;


		this->organizmy = new Organizm * [this->Getwyskosc() * this->Getszerokosc()];

		for (int i = 0; i < this->Getwyskosc() * this->Getszerokosc(); i++)
		{
			organizmy[i] = nullptr;
		}

		int id = 0, x = 0, y = 0, sila = 0, rundy = 0;
		bool dz = 0;
		Organizm* org = nullptr;

		//fopen_s(&plik, "org.bin", "r");
		std::ifstream infile("output.txt");
		if (infile)
		{

			for (int i = 0; i < size; i++)
			{
				/*
				fread(&id, sizeof(int), 1, plik);
				fread(&x, sizeof(int), 1, plik);
				fread(&y, sizeof(int), 1, plik);
				fread(&sila, sizeof(int), 1, plik);
				fread(&rundy, sizeof(int), 1, plik);
				fread(&dz, sizeof(bool), 1, plik);
				*/
				infile >> id >> x >> y >> sila >> rundy >> dz;
				if (id == IDCZLOWIEK)
					org = new Czlowiek(x, y, dz, this);
				else if (id == IDWILK)
					org = new Wilk(x, y, dz, this);
				else if (id == IDOWCA)
					org = new Owca(x, y, dz, this);
				else if (id == IDLIS)
					org = new Lis(x, y, dz, this);
				else if (id == IDZOLW)
					org = new Zolw(x, y, dz, this);
				else if (id == IDANTYLOPA)
					org = new Antylopa(x, y, dz, this);
				else if (id == IDTRAWA)
					org = new Trawa(x, y, dz, this);
				else if (id == IDMLECZ)
					org = new Mlecz(x, y, dz, this);
				else if (id == IDGUARANA)
					org = new Guarana(x, y, dz, this);
				else if (id == IDWILCZEJAGODY)
					org = new WilczeJagody(x, y, dz, this);
				else if (id == IDBARSZCZSOS)
					org = new BarszczSosnowskiego(x, y, dz, this);

				if (org != nullptr)
				{
					org->SetSila(sila);
					org->SetRundy(rundy);
				}
			}
			infile.close();
			//fclose(plik);
		}
	}
	else
	{
		this->komunikat = "Nie ma save";
	}
}

Swiat::~Swiat()
{
	if (organizmy != nullptr)
	{
		for (int i = 0; i < szerokosc * wysokosc; i++)
		{
			if (organizmy[i] != nullptr)
				delete organizmy[i];
		}
		delete[] organizmy;
	}
	for (int i = 0; i < ILIST; i++)
		this->kolejnosc[i].clear();
}

File: Python/Swiat.py
import random
import csv
from Czlowiek import Czlowiek
from Antylopa import Antylopa
from BarszczSosnowskiego import BarszczSosnowskiego
from Guarana import Guarana
from Lis import Lis
from Mlecz import Mlecz
from Owca import Owca
from Trawa import Trawa
from WilczeJagody import WilczeJagody
from Wilk import Wilk
from Zolw import Zolw
from CyberOwca import CyberOwca


class Swiat:
    def __init__(self, m=40, n=40, uklad_planszy=4):
        self.GORA = 0
        self.DOL = 1
        self.PRAWO = 2
        self.LEWO = 3
        self.LEWOG = 4
        self.PRAWOG = 5
        self.BRAK = 6
        self.ILIST = 5
        self.CZAS_UMIEJETNOSCI = 5
        self.CZAS_ROZMNAZANIA = 2

        self.ROZMNAZANIE = 1
        self.ROZPRZESTRZENIANIE = 2
        self.KILL = 3
        self.ODBITY = 4
        self.UCIEKA = 5
        self.UMIEJETNOSC_A = 6
        self.UMIEJETNOSC_N = 7

        self.IDCZLOWIEK = 1
        self.IDWILK = 2
        self.IDOWCA = 3
        self.IDLIS = 4
        self.IDZOLW = 5
        self.IDANTYLOPA = 6
        self.IDTRAWA = 7
        self.IDMLECZ = 8
        self.IDGUARANA = 9
        self.IDWILCZEJAGODY = 10
        self.IDBARSZCZSOS = 11
        self.IDCYBEROWCA = 12

        self.szerokosc = m
        self.wysokosc = n
        self.uklad_planszy = uklad_planszy
        self.kolejnosc = [[] for _ in range(self.ILIST)]
        self.wydarzenia = []
        self.organizmy = [None] * (m * n)
        self.kierunek_ruchu = self.BRAK
        self.rundy = self.CZAS_UMIEJETNOSCI
        self.aktywna = False
        self.czlowiekzyje = True
        self.komunikat = ""
        self.rand = random.Random()

        self.x_barszczu = -1
        self.y_barszczu = -1

        org = Czlowiek(0, 0, self)
        for _ in range((m * n) // 10):
            id = self.rand.randint(2, 12)
            x = self.rand.randint(0, m - 1)
            y = self.rand.randint(0, n - 1)
            if self.GetOrganizmy(x, y) is None:
                if id == self.IDWILK:
                    org = Wilk(x, y, False, self)
                elif id == self.IDOWCA:
                    org = Owca(x, y, False, self)
                elif id == self.IDLIS:
                    org = Lis(x, y, False, self)
                elif id == self.IDZOLW:
                    org = Zolw(x, y, False, self)
                elif id == self.IDANTYLOPA:
                    org = Antylopa(x, y, False, self)
                elif id == self.IDTRAWA:
                    org = Trawa(x, y, False, self)
                elif id == self.IDMLECZ:
                    org = Mlecz(x, y, False, self)
                elif id == self.IDGUARANA:
                    org = Guarana(x, y, False, self)
                elif id == self.IDWILCZEJAGODY:
                    org = WilczeJagody(x, y, False, self)
                elif id == self.IDBARSZCZSOS:
                    org = BarszczSosnowskiego(x, y, False, self)
                elif id == self.IDCYBEROWCA:
                    org = CyberOwca(x, y, False, self)

    def wykonajTure(self):
        self.znajdz_Barszcz()
        for i in range(self.ILIST - 1, -1, -1):
            j = 0
            while j < len(self.kolejnosc[i]):
                if not self.kolejnosc[i][j].GetDZ():
                    self.kolejnosc[i][j].KolejnaRunda()
                    self.kolejnosc[i][j].akcja()
                else:
                    self.kolejnosc[i][j].SetDZ(False)
                j += 1
        if self.rundy >= self.CZAS_UMIEJETNOSCI and self.aktywna:
            self.ZmienAktywna()
            self.WyzerujRundy()
            self.dodajWydarzenie(None, None, self.UMIEJETNOSC_N)
        self.KolejnaRunda()

    def umiejetnosc(self):
        if not self.czlowiekzyje:
            self.komunikat = "Nie można użyć umiejętności, człowiek nie żyje"
        elif self.rundy >= self.CZAS_UMIEJETNOSCI:
            self.ZmienAktywna()
            self.WyzerujRundy()
            if self.aktywna:
                self.dodajWydarzenie(None, None, self.UMIEJETNOSC_A)
            else:
                self.dodajWydarzenie(None, None, self.UMIEJETNOSC_N)
        else:
            if self.CzyAktywna():
                self.komunikat = f"Umiejętność aktywna przez {self.CZAS_UMIEJETNOSCI - self.rundy} "
                if self.CZAS_UMIEJETNOSCI - self.rundy == 1:
                    self.komunikat += "turę"
                elif self.CZAS_UMIEJETNOSCI - self.rundy == 5:
                    self.komunikat += "tur"
                else:
                    self.komunikat += "tur"
            else:
                self.komunikat = f"Umiejętność gotowa za {self.CZAS_UMIEJETNOSCI - self.rundy} "
                if self.CZAS_UMIEJETNOSCI - self.rundy == 1:
                    self.komunikat += "turę"
                elif self.CZAS_UMIEJETNOSCI - self.rundy == 5:
                    self.komunikat += "tur"
                else:
                    self.komunikat += "turę"

    def rysujSwiat(self, frame):
        frame.updateButtons(self)

    def addToList(self, organizm):
        i = organizm.GetInicjatywa()
        if i < 2:
            self.kolejnosc[i].append(organizm)
        elif i == 5:
            self.kolejnosc[3].append(organizm)
        elif i == 7:
            self.kolejnosc[4].append(organizm)
        else:
            self.kolejnosc[2].append(organizm)

    def deletefromList(self, organizm):
        i = organizm.GetInicjatywa()
        j = 0

        if isinstance(organizm, Czlowiek):
            self.czlowiekzyje = False

        if i < 2:
            while j < len(self.kolejnosc[i]):
                if self.kolejnosc[i][j] != organizm:
                    j += 1
                else:
                    self.kolejnosc[i].pop(j)
        elif i == 5:
            while j < len(self.kolejnosc[3]):
                if self.kolejnosc[3][j] != organizm:
                    j += 1
                else:
                    self.kolejnosc[3].pop(j)
        elif i == 7:
            while j < len(self.kolejnosc[4]):
                if self.kolejnosc[4][j] != organizm:
                    j += 1
                else:
                    self.kolejnosc[4].pop(j)
        else:
            while j < len(self.kolejnosc[2]):
                if self.kolejnosc[2][j] != organizm:
                    j += 1
                else:
                    self.kolejnosc[2].pop(j)

    def Getszerokosc(self):
        return self.szerokosc

    def Getwyskosc(self):
        return self.wysokosc

    def Setkierunek(self, kierunek):
        self.kierunek_ruchu = kierunek

    def Getkierunek(self):
        return self.kierunek_ruchu

    def Getuklad_planszy(self):
        return self.uklad_planszy

    def SetOrganizmy(self, x, y, n):
        self.organizmy[y * self.Getszerokosc() + x] = n

    def GetOrganizmy(self, x, y):
        return self.organizmy[y * self.Getszerokosc() + x]

    def WyzerujRundy(self):
        self.rundy = 0

    def KolejnaRunda(self):
        self.rundy += 1

    def GetRundy(self):
        return self.rundy

    def GetKomunikat(self):
        return self.komunikat

    def SetRundy(self, r):
        self.rundy = r

    def CzyAktywna(self):
        return self.aktywna

    def ZmienAktywna(self):
        if self.aktywna:
            self.aktywna = False
        else:
            self.aktywna = True

    def dodajWydarzenie(self, organizm1, organizm2, kod):
        s = ""

        if organizm1 is not None:
            s += organizm1.UczestnikWydarzenia()

        if kod == self.ROZMNAZANIE:
            s += " rozmnazyl sie z "
        elif kod == self.ROZPRZESTRZENIANIE:
            s += " rozprzestrzenil sie"
        elif kod == self.KILL:
            s += " zjadl "
        elif kod == self.ODBITY:
            s += " odbil atak "
        elif kod == self.UCIEKA:
            s += " ucieka od "
        elif kod == self.UMIEJETNOSC_A:
            s += "Czlowiek uzyl umiejetnosci"
        elif kod == self.UMIEJETNOSC_N:
            s += "Umiejetnosc przestala dzialac"

        if organizm2 is not None:
            s += organizm2.UczestnikWydarzenia()

        self.wydarzenia.append(s)

    def save(self):

        size = 0
        for i in range(self.ILIST):
            size += len(self.kolejnosc[i])

        rows = [
            [self.szerokosc, self.wysokosc, self.kierunek_ruchu, self.rundy, self.uklad_planszy, self.aktywna, self.czlowiekzyje, size]
        ]
        with open("plik.csv", mode="w", newline="") as file:
            writer = csv.writer(file)  # Utworzenie obiektu writer

            # Zapisanie wierszy danych
            for row in rows:
                writer.writerow(row)

        with open("file.csv", mode="w", newline="") as file:
            writer = csv.writer(file)  # Create a writer object

            # Loop to add rows
            for i in range(self.ILIST - 1, -1, -1):
                j = 0
                while j < len(self.kolejnosc[i]):
                    row = [self.kolejnosc[i][j].GetId(), self.kolejnosc[i][j].GetX(), self.kolejnosc[i][j].GetY(), self.kolejnosc[i][j].GetSila(), self.kolejnosc[i][j].GetRundy(), self.kolejnosc[i][j].GetDZ()]

                    # Write the row to the CSV file
                    writer.writerow(row)
                    j += 1
    def load(self, gra):
        with open("plik.csv", mode="r") as file:
            reader = csv.reader(file)  # Utworzenie obiektu reader

            # Odczytanie wierszy danych
            for row in reader:
                self.szerokosc = int(row[0])
                self.wysokosc = int(row[1])
                self.kierunek_ruchu = int(row[2])
                self.rundy = int(row[3])
                self.uklad_planszy = int(row[4])
                self.aktywna = bool(row[5])
                self.czlowiekzyje = bool(row[6])
                size = int(row[7])

            self.organizmy = None

            self.organizmy = [None] * (self.szerokosc * self.wysokosc)

            for i in range(self.ILIST):
                self.kolejnosc[i].clear()

        with open("file.csv", mode="r") as file:
            reader = csv.reader(file)
            for row in reader:

                org = None

                id = int(row[0])
                x = int(row[1])
                y = int(row[2])
                sila = int(row[3])
                rundy = int(row[4])
                dz = bool(row[5])
                if id == self.IDCZLOWIEK:
                    org = Czlowiek(x, y, self)
                elif id == self.IDWILK:
                    org = Wilk(x, y, dz, self)
                elif id == self.IDOWCA:
                    org = Owca(x, y, dz, self)
                elif id == self.IDLIS:
                    org = Lis(x, y, dz, self)
                elif id == self.IDZOLW:
                    org = Zolw(x, y, dz, self)
                elif id == self.IDANTYLOPA:
                    org = Antylopa(x, y, dz, self)
                elif id == self.IDTRAWA:
                    org = Trawa(x, y, dz, self)
                elif id == self.IDMLECZ:
                    org = Mlecz(x, y, dz, self)
                elif id == self.IDGUARANA:
                    org = Guarana(x, y, dz, self)
                elif id == self.IDWILCZEJAGODY:
                    org = WilczeJagody(x, y, dz, self)
                elif id == self.IDBARSZCZSOS:
                    org = BarszczSosnowskiego(x, y, dz, self)

                if org is not None:
                    org.SetSila(sila)
                    org.SetRundy(rundy)

        gra.Load(self)

    def CzypusteWydarzenia(self):
        return len(self.wydarzenia) == 0

    def wezWydarzenie(self):
        r = self.wydarzenia[0]
        self.wydarzenia.pop(0)
        return r

    def Usunkomunikat(self):
        self.komunikat = ""

    def znajdz_Barszcz(self):
        if(self.x_barszczu == -1):
            j = 0
            istnieje = False
            while j < len(self.kolejnosc[0]):
                if (self.kolejnosc[0][j].GetId() == self.IDBARSZCZSOS):
                    self.x_barszczu = self.kolejnosc[0][j].GetX()
                    self.y_barszczu = self.kolejnosc[0][j].GetY()
                    istnieje = True
                    break
                j += 1
            if(not istnieje):
                self.x_barszczu = -1
                self.y_barszczu = -1
        elif(self.GetOrganizmy(self.x_barszczu, self.y_barszczu).GetId() != self.IDBARSZCZSOS):
            j = 0
            istnieje = False
            while j < len(self.kolejnosc[0]):
                if(self.kolejnosc[0][j].GetId() == self.IDBARSZCZSOS):
                    self.x_barszczu = self.kolejnosc[0][j].GetX()
                    self.y_barszczu = self.kolejnosc[0][j].GetY()
                    istnieje = True
                    break
                j += 1
            if (not istnieje):
                self.x_barszczu = -1
                self.y_barszczu = -1